### PR TITLE
msg/async: support of non-block connect in async messenger

### PR DESCRIPTION
**Pull request by zuiwanyuan**
msg/async: support of non-block connect in async messenger

Fixes: #12802

Signed-off-by: Jianhui Yuan <zuiwanyuan@gmail.com>

### CODE REVIEW

File: src/msg/async/AsyncConnection.cc [hunk 1]
```diff
@@ -971,16 +971,27 @@ int AsyncConnection::_process_connection()
...
+        break;
+      }
+
+    case STATE_CONNECTING_RE:
+      {
+        r = net.reconnect(get_peer_addr(), sd);
         if (r < 0) {
+          if (r == -EINPROGRESS || r == -EALREADY)
+            break;
           goto fail;
```
Comment (yuyuyu101) [inline src/msg/async/AsyncConnection.cc]: plz add level 1 error log above "goto"

File: src/msg/async/AsyncConnection.cc [hunk 2]
```diff
@@ -2078,7 +2089,7 @@ void AsyncConnection::fault()
   }
 
   write_lock.Lock();
-  if (sd >= 0) {
+  if (sd >= 0 && state != STATE_CONNECTING_RE) {
```
Comment (yuyuyu101) [inline src/msg/async/AsyncConnection.cc]: why not?

File: src/msg/async/AsyncConnection.cc [hunk 3]
```diff
@@ -2132,7 +2143,9 @@ void AsyncConnection::fault()
       if (backoff > async_msgr->cct->_conf->ms_max_backoff)
         backoff.set_from_double(async_msgr->cct->_conf->ms_max_backoff);
     }
-    state = STATE_CONNECTING;
+
+    if (state != STATE_CONNECTING_RE)
+      state = STATE_CONNECTING;
```
Comment (yuyuyu101) [inline src/msg/async/AsyncConnection.cc]: need this?

File: src/msg/async/net_handler.cc [hunk 4]
```diff
@@ -132,6 +132,18 @@ int NetHandler::generic_connect(const entity_addr_t& addr, bool nonblock)
   return s;
 }
 
+int NetHandler::reconnect(const entity_addr_t &addr, int sd)
+{
+  int ret = ::connect(sd, (sockaddr*)&addr.addr, addr.addr_size());
+
+  if (ret < 0 && errno != EISCONN) {
```
Comment (yuyuyu101) [inline src/msg/async/net_handler.cc]: let caller judge or we need to hidden EINPROGRESS and EALREADY

File: src/msg/async/AsyncConnection.cc (hunk 1 above)
Comment (zuiwanyuan) [inline src/msg/async/AsyncConnection.cc]: OK

File: src/msg/async/AsyncConnection.cc (hunk 2 above)
Comment (zuiwanyuan) [inline src/msg/async/AsyncConnection.cc]: Oh. We need to reduce it. When we goto fail, we just convert to STATE_CONNECTING.

File: src/msg/async/net_handler.cc (hunk 4 above)
Comment (zuiwanyuan) [inline src/msg/async/net_handler.cc]: Let us hidden EINPROGRESS and EALREADY.

**Comment by yuyuyu101** [comment]
LGTM

**Comment by yuyuyu101** [comment]
Tested in http://pulpito.ceph.com/haomai-2015-11-10_11:24:27-rados-wip-haomai-testing-distro-basic-multi/.

Failed jobs 1145091 and 1145178 his http://tracker.ceph.com/issues/13767
Failed job 1145207 hits fd limits

**Comment by yuyuyu101** [comment]
ready to merge

**Comment by yuyuyu101** [comment]
@liewegas give a look? It has passed tests.

File: src/msg/async/net_handler.cc [hunk 5]
```diff
@@ -132,6 +132,26 @@ int NetHandler::generic_connect(const entity_addr_t& addr, bool nonblock)
   return s;
 }
 
+/* Try to reconnect the socket.
+ *
+ * @return    0         success
+ *            > 0       just break, and wait for event
+ *            < 0       need to goto fail
+ */
+int NetHandler::reconnect(const entity_addr_t &addr, int sd)
```
Comment (tchaikov) [inline src/msg/async/net_handler.cc]: doxygen comments are normally put into the header file. and should be started with `/**`, see src/mon/Monitor.h for more examples.

Comment (zuiwanyuan) [inline src/msg/async/net_handler.cc]: ok

**Pull request merged**